File: cli/src/commands/update.js
import path from 'path';
import fs from 'fs/promises';
import { loadGraph, loadMeta, saveGraph, computeFileHash } from '../graph.js';
import { traverseFiles, detectLanguage } from '../traverser.js';
import { initParser, parseFile } from '../parser.js';
import { detectModuleName } from '../scanner.js';
import { detectChangedFiles, mergeGraphUpdate } from '../differ.js';
import { saveSlices } from '../slicer.js';

/**
 * Try to get the current git commit hash from the given directory.
 * Returns null if the directory is not a git repo or simple-git is unavailable.
 *
 * @param {string} dir - Directory to check for git.
 * @returns {Promise<string|null>} The current commit hash or null.
 */
async function getGitCommitHash(dir) {
  try {
    const { simpleGit } = await import('simple-git');
    const git = simpleGit(dir);
    const isRepo = await git.checkIsRepo();
    if (!isRepo) return null;
    const log = await git.log({ maxCount: 1 });
    return log.latest ? log.latest.hash : null;
  } catch {
    return null;
  }
}

/**
 * Register the `update` command on the given Commander program.
 *
 * @param {import('commander').Command} program - The Commander program instance.
 */
export function registerUpdateCommand(program) {
  program
    .command('update [dir]')
    .description('Incrementally update the code graph for changed files')
    .action(async (dir) => {
      const rootDir = path.resolve(dir || '.');
      const outputDir = path.join(rootDir, '.codemap');

      // Step 1: Load existing graph and meta
      let graph, meta;
      try {
        graph = await loadGraph(outputDir);
        meta = await loadMeta(outputDir);
      } catch {
        console.error('No existing code graph found. Run "codegraph scan" first.');
        process.exit(1);
      }

      const startTime = Date.now();

      // Step 2: Initialize parser
      await initParser();

      // Step 3: Traverse files and compute current hashes
      const files = await traverseFiles(rootDir);
      const currentHashes = {};

      for (const absPath of files) {
        const language = detectLanguage(absPath);
        if (!language) continue;

        const relPath = path.relative(rootDir, absPath).replace(/\\/g, '/');
        try {
          const content = await fs.readFile(absPath, 'utf-8');
          currentHashes[relPath] = computeFileHash(content);
        } catch {
          // Skip files that can't be read
          continue;
        }
      }

      // Step 4: Detect changes
      const changes = detectChangedFiles(meta.fileHashes || {}, currentHashes);

      if (changes.added.length === 0 && changes.modified.length === 0 && changes.removed.length === 0) {
        console.log('No changes detected.');
        return;
      }

      // Step 5: Re-parse only changed/added files
      const updatedFiles = {};
      const changedPaths = [...changes.added, ...changes.modified];

      for (const relPath of changedPaths) {
        const absPath = path.resolve(rootDir, relPath);
        const language = detectLanguage(absPath);
        if (!language) continue;

        let parsed;
        try {
          parsed = await parseFile(absPath, language);
        } catch {
          continue;
        }

        const moduleName = detectModuleName(absPath, rootDir);
        const content = await fs.readFile(absPath, 'utf-8');
        const hash = computeFileHash(content);
        const baseName = path.basename(absPath, path.extname(absPath)).toLowerCase();
        const isEntryPoint = ['main', 'index', 'server', 'app', 'entry', 'bootstrap'].includes(baseName);

        updatedFiles[relPath] = {
          language,
          module: moduleName,
          hash,
          lines: parsed.lines,
          functions: parsed.functions,
          classes: parsed.classes,
          types: parsed.types,
          imports: parsed.imports,
          exports: parsed.exports,
          isEntryPoint,
        };
      }

      // Step 6: Merge updates into graph
      mergeGraphUpdate(graph, updatedFiles, changes.removed);

      // Step 7: Update metadata
      const commitHash = await getGitCommitHash(rootDir);
      graph.commitHash = commitHash;
      graph.scannedAt = new Date().toISOString();

      const scanDuration = Date.now() - startTime;

      // Build updated file hashes
      const fileHashes = {};
      for (const [filePath, fileData] of Object.entries(graph.files)) {
        fileHashes[filePath] = fileData.hash;
      }

      meta.lastScanAt = new Date().toISOString();
      meta.commitHash = commitHash;
      meta.scanDuration = scanDuration;
      meta.fileHashes = fileHashes;

      // Step 8: Save graph + meta + re-generate slices
      await saveGraph(outputDir, graph, meta);
      await saveSlices(outputDir, graph);

      // Step 9: Print change summary
      console.log(`Update complete.`);
      console.log(`  +${changes.added.length} ~${changes.modified.length} -${changes.removed.length}`);
      if (changes.added.length > 0) {
        console.log(`  Added: ${changes.added.join(', ')}`);
      }
      if (changes.modified.length > 0) {
        console.log(`  Modified: ${changes.modified.join(', ')}`);
      }
      if (changes.removed.length > 0) {
        console.log(`  Removed: ${changes.removed.join(', ')}`);
      }
    });
}

File: rust-cli/src/commands/update.rs
use clap::Args;
use std::collections::HashMap;
use std::path::PathBuf;

#[derive(Args)]
pub struct UpdateArgs {
    /// Project directory
    pub dir: Option<String>,
    /// Additional glob patterns to exclude
    #[arg(long, num_args = 1..)]
    pub exclude: Vec<String>,
}

pub fn run(args: UpdateArgs) {
    let dir = args.dir.unwrap_or_else(|| ".".to_string());
    let root = PathBuf::from(&dir);
    let root = match root.canonicalize() {
        Ok(p) => p,
        Err(e) => {
            eprintln!("Error: cannot resolve directory '{}': {}", dir, e);
            std::process::exit(1);
        }
    };

    let codemap_dir = root.join(".codemap");

    // 加载现有图谱
    let mut graph = match crate::graph::load_graph(&codemap_dir) {
        Ok(g) => g,
        Err(e) => {
            eprintln!(
                "Error: could not load graph from {}/.codemap/: {}",
                root.display(),
                e
            );
            eprintln!("Run 'codegraph scan {}' first.", root.display());
            std::process::exit(1);
        }
    };

    // 遍历磁盘当前文件，计算哈希
    let files = crate::traverser::traverse_files(&root, &args.exclude);
    let has_cpp = crate::traverser::has_cpp_source_files(&files);

    let mut new_hashes: HashMap<String, String> = HashMap::new();
    let mut file_contents: HashMap<String, Vec<u8>> = HashMap::new();

    for abs_path in &files {
        if crate::traverser::detect_language(abs_path).is_none() {
            continue;
        }
        let content = match std::fs::read(abs_path) {
            Ok(c) => c,
            Err(_) => continue,
        };

        let rel_path = abs_path
            .strip_prefix(&root)
            .unwrap_or(abs_path)
            .to_string_lossy()
            .replace('\\', "/");

        let hash = crate::graph::compute_file_hash(&content);
        new_hashes.insert(rel_path.clone(), hash);
        file_contents.insert(rel_path, content);
    }

    // 从 meta.fileHashes 读取旧哈希（与 Node.js update 逻辑一致）
    // 若 meta.json 不存在或无 fileHashes，回退到从 graph.files 提取
    let old_hashes: HashMap<String, String> =
        match crate::graph::load_meta(&codemap_dir) {
            Ok(meta) if !meta.file_hashes.is_empty() => meta.file_hashes.into_iter().collect(),
            _ => graph
                .files
                .iter()
                .map(|(p, f)| (p.clone(), f.hash.clone()))
                .collect(),
        };

    // 检测变更
    let changes = crate::differ::detect_changed_files(&old_hashes, &new_hashes);

    if changes.is_empty() {
        println!("No changes detected.");
        return;
    }

    println!(
        "Changes: +{} added, ~{} modified, -{} removed",
        changes.added.len(),
        changes.modified.len(),
        changes.removed.len()
    );

    // 解析变更文件（新增 + 修改）
    let mut updated_files: HashMap<String, crate::graph::FileEntry> = HashMap::new();

    for rel_path in changes.added.iter().chain(changes.modified.iter()) {
        let content = match file_contents.get(rel_path) {
            Some(c) => c,
            None => continue,
        };

        // 重建绝对路径以检测语言
        let abs_path = root.join(rel_path.replace('/', std::path::MAIN_SEPARATOR_STR));
        let base_lang = match crate::traverser::detect_language(&abs_path) {
            Some(l) => l,
            None => continue,
        };
        let lang = crate::traverser::effective_language(&abs_path, base_lang, has_cpp);

        let adapter = crate::languages::get_adapter(lang);

        let mut ts_parser = tree_sitter::Parser::new();
        ts_parser.set_language(&adapter.language()).ok();
        let tree = match ts_parser.parse(content, None) {
            Some(t) => t,
            None => continue,
        };

        let lang_functions = adapter.extract_functions(&tree, content);
        let lang_imports = adapter.extract_imports(&tree, content);
        let lang_exports = adapter.extract_exports(&tree, content);
        let lang_classes = adapter.extract_classes(&tree, content);
        let lines = content.iter().filter(|&&b| b == b'\n').count() as u32 + 1;

        let functions = crate::scanner::convert_functions(&lang_functions);
        let classes = crate::scanner::convert_classes(&lang_classes);
        let types = crate::scanner::convert_types(&lang_classes, lang);
        let imports = crate::scanner::convert_imports(&lang_imports);
        let exports = crate::scanner::convert_exports(&lang_exports);

        let module_name = crate::scanner::detect_module_name(&abs_path, &root);
        let hash = new_hashes[rel_path].clone();

        updated_files.insert(
            rel_path.clone(),
            crate::graph::FileEntry {
                language: lang.as_str().to_string(),
                module: module_name,
                hash,
                lines,
                functions,
                classes,
                types,
                imports,
                exports,
                is_entry_point: crate::graph::is_entry_point(&abs_path),
            },
        );
    }

    // 合并变更到图谱
    crate::differ::merge_graph_update(&mut graph, updated_files, &changes.removed);

    // 更新扫描时间
    graph.scanned_at = crate::graph::chrono_now();

    // 保存更新后的图谱
    if let Err(e) = crate::graph::save_graph(&codemap_dir, &graph) {
        eprintln!("Error saving graph: {}", e);
        std::process::exit(1);
    }

    // 重新生成 slices（与 Node.js update 行为一致）
    if let Err(e) = crate::slicer::save_slices(&codemap_dir, &graph) {
        eprintln!("Warning: failed to save slices: {}", e);
    }

    println!("Update complete.");
    println!(
        "  +{} ~{} -{}",
        changes.added.len(),
        changes.modified.len(),
        changes.removed.len()
    );
    if !changes.added.is_empty() {
        println!("  Added: {}", changes.added.join(", "));
    }
    if !changes.modified.is_empty() {
        println!("  Modified: {}", changes.modified.join(", "));
    }
    if !changes.removed.is_empty() {
        println!("  Removed: {}", changes.removed.join(", "));
    }
}
